File: docs/test-cases/csrf.java
// Test case: csrf (A01:2025)
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.GetMapping;

@Configuration
@EnableWebSecurity
public class SecurityConfig extends WebSecurityConfigurerAdapter {
    @Override
    protected void configure(HttpSecurity http) throws Exception {
        // BUG: CSRF protection explicitly disabled
        http.csrf().disable()
            .authorizeRequests()
            .anyRequest().authenticated();
    }
}

@Controller
class TransferController {
    @GetMapping("/transfer")
    public String showForm() {
        // BUG: form template has no _csrf hidden field
        return "transfer-form";
        // transfer-form.html contains:
        // <form method="POST" action="/transfer">
        //   <input name="amount"/>
        //   <input name="to"/>
        //   <button>Send</button>
        // </form>
    }

    @PostMapping("/transfer")
    public String doTransfer(String amount, String to) {
        // BUG: no CSRF token validated — disabled globally above
        bankService.transfer(amount, to);
        return "redirect:/done";
    }
}

File: docs/test-cases/csrf.rs
// Test case: csrf (A01:2025)
use actix_web::{web, App, HttpServer, HttpRequest, HttpResponse};

// BUG: no CSRF middleware registered on the application
#[actix_web::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| {
        App::new()
            .route("/transfer", web::get().to(show_form))
            .route("/transfer", web::post().to(do_transfer))
    })
    .bind("127.0.0.1:8080")?
    .run()
    .await
}

async fn show_form() -> HttpResponse {
    // BUG: form has no CSRF token hidden field
    HttpResponse::Ok().content_type("text/html").body(r#"
        <form method="POST" action="/transfer">
            <input name="amount"/>
            <input name="to"/>
            <button>Send</button>
        </form>
    "#)
}

// BUG: state-changing endpoint accepts POST with no CSRF token validation
async fn do_transfer(req: HttpRequest, form: web::Form<TransferForm>) -> HttpResponse {
    execute_transfer(&form.amount, &form.to);
    HttpResponse::SeeOther()
        .insert_header(("Location", "/done"))
        .finish()
}

#[derive(serde::Deserialize)]
struct TransferForm {
    amount: String,
    to: String,
}

fn execute_transfer(amount: &str, to: &str) {}
